### DISCUSSION TITLE
milvus null and default value functionality is not working

### DISCUSSION THREAD

**Discussion by vihariazure** [milvus-io/milvus#38346] [Q&A and General discussion]
https://milvus.io/docs/nullable-and-default.md

hi team i tired with above code regarding using none values 

from pymilvus import MilvusClient, DataType

client = MilvusClient(uri='http://localhost:19530')

schema = client.create_schema()

schema.add_field(field_name="id", datatype=DataType.INT64, is_primary=True)
schema.add_field(field_name="vector", datatype=DataType.FLOAT_VECTOR, dim=5)
schema.add_field(field_name="age", datatype=DataType.INT64, nullable=True) # Nullable field

index_params = client.prepare_index_params()
index_params.add_index(field_name="vector", index_type="IVF_FLAT", metric_type="L2", params={ "nlist": 128 })

client.create_collection(collection_name="user_profiles_null", schema=schema, index_params=index_params)

data = [
    {"id": 1, "vector": [0.1, 0.2, 0.3, 0.4, 0.5], "age": 30},
    {"id": 2, "vector": [0.2, 0.3, 0.4, 0.5, 0.6], "age": None},
]

client.insert(collection_name="user_profiles_null", data=data)

below is the error

File d:\milvus_connect\milvus\lib\site-packages\pymilvus\client\entity_helper.py:264, in pack_field_value_to_field_data(field_value, field_data, field_info)
    [262](file:///D:/milvus_connect/milvus/lib/site-packages/pymilvus/client/entity_helper.py:262)         field_data.scalars.long_data.data.append(field_value)
    [263](file:///D:/milvus_connect/milvus/lib/site-packages/pymilvus/client/entity_helper.py:263)     except (TypeError, ValueError) as e:
--> [264](file:///D:/milvus_connect/milvus/lib/site-packages/pymilvus/client/entity_helper.py:264)         raise DataNotMatchException(
    [265](file:///D:/milvus_connect/milvus/lib/site-packages/pymilvus/client/entity_helper.py:265)             message=ExceptionsMessage.FieldDataInconsistent
    [266](file:///D:/milvus_connect/milvus/lib/site-packages/pymilvus/client/entity_helper.py:266)             % (field_name, "int64", type(field_value))
    [267](file:///D:/milvus_connect/milvus/lib/site-packages/pymilvus/client/entity_helper.py:267)         ) from e
    [268](file:///D:/milvus_connect/milvus/lib/site-packages/pymilvus/client/entity_helper.py:268) elif field_type == DataType.FLOAT:
    [269](file:///D:/milvus_connect/milvus/lib/site-packages/pymilvus/client/entity_helper.py:269)     try:

**DataNotMatchException: <DataNotMatchException: (code=1, message=The Input data type is inconsistent with defined schema, {age} field should be a int64, but got a {<class 'NoneType'>} instead.)>**

**Comment by yhmo** [comment]
I guess your milvus server is not v2.5.0.
Use this pymilvus method to print the server version:
`print(client.get_server_version())`

**Reply by yhmo** [reply]
Use this script to test:

```
from pymilvus import MilvusClient, DataType

client = MilvusClient(uri='http://localhost:19530')

print(client.get_server_version())

schema = client.create_schema()

schema.add_field(field_name="id", datatype=DataType.INT64, is_primary=True)
schema.add_field(field_name="vector", datatype=DataType.FLOAT_VECTOR, dim=5)
schema.add_field(field_name="age", datatype=DataType.INT64, nullable=True) # Nullable field

index_params = client.prepare_index_params()
index_params.add_index(field_name="vector", index_type="IVF_FLAT", metric_type="L2", params={ "nlist": 128 })

client.drop_collection(collection_name="user_profiles_null")
client.create_collection(collection_name="user_profiles_null", schema=schema, index_params=index_params)

data = [
{"id": 1, "vector": [0.1, 0.2, 0.3, 0.4, 0.5], "age": 30},
{"id": 2, "vector": [0.2, 0.3, 0.4, 0.5, 0.6], "age": None},
]

client.insert(collection_name="user_profiles_null", data=data)

results = client.query(collection_name="user_profiles_null",
                       filter="id>0",
                       consistency_level="Strong",
                       output_fields=["*"])
for res in results:
    print(res)
```

If your milvus server is v2.5, it will print these lines:
```
v2.5.0-beta
{'vector': [0.1, 0.2, 0.3, 0.4, 0.5], 'age': 30, 'id': 1}
{'vector': [0.2, 0.3, 0.4, 0.5, 0.6], 'age': None, 'id': 2}

Process finished with exit code 0
```

**Reply by vihariazure** [reply]
from pymilvus import MilvusClient, DataType

client = MilvusClient(uri='http://localhost:19530')

print(client.get_server_version())

{
	"name": "AttributeError",
	"message": "'MilvusClient' object has no attribute 'get_server_version'",
	"stack": "---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[6], line 5
      1 from pymilvus import MilvusClient, DataType
      3 client = MilvusClient(uri='http://localhost:19530')
----> 5 print(client.get_server_version())

**AttributeError: 'MilvusClient' object has no attribute 'get_server_version'"**
}

from pymilvus import connections, utility

connections.connect()

server_version = utility.get_server_version()
print(server_version)

**v2.4.15**

**Reply by yhmo** [reply]
Just double-check your milvus server version.
2.4.x doesn't support null/default value.

**Reply by vihariazure** [reply]
**still the same error even after upgrading milvus image**

from pymilvus import connections, utility

connections.connect()

server_version = utility.get_server_version()
print(server_version)

**v2.5.0-beta**

upgraded the milvus image 

from pymilvus import MilvusClient, DataType

client = MilvusClient(uri='http://localhost:19530')

# print(client.get_server_version())

schema = client.create_schema()

schema.add_field(field_name="id", datatype=DataType.INT64, is_primary=True)
schema.add_field(field_name="vector", datatype=DataType.FLOAT_VECTOR, dim=5)
schema.add_field(field_name="age", datatype=DataType.INT64, nullable=True) # Nullable field

index_params = client.prepare_index_params()
index_params.add_index(field_name="vector", index_type="IVF_FLAT", metric_type="L2", params={ "nlist": 128 })

client.drop_collection(collection_name="user_profiles_null")
client.create_collection(collection_name="user_profiles_null", schema=schema, index_params=index_params)

data = [
{"id": 1, "vector": [0.1, 0.2, 0.3, 0.4, 0.5], "age": 30},
{"id": 2, "vector": [0.2, 0.3, 0.4, 0.5, 0.6], "age": None},
]

client.insert(collection_name="user_profiles_null", data=data)

results = client.query(collection_name="user_profiles_null",
                       filter="id>0",
                       consistency_level="Strong",
                       output_fields=["*"])
for res in results:
    print(res)
    
    
    **DataNotMatchException: <DataNotMatchException: (code=1, message=The Input data type is inconsistent with defined schema, {age} field should be a int64, but got a {<class 'NoneType'>} instead.)>**

**Reply by yhmo** [reply]
Null/Default is a new feature of v2.5. **Both milvus version and pymilvus version need to be upgraded to v2.5.x**

```
pip3 uninstall pymilvus
pip3 install pymilvus==2.5.0
```
